### PR TITLE
Bump golang to 1.23.1

### PR DESCRIPTION
**Pull request by huww98**
external-snapshotter is already using 1.23.1 in go.mod

```release-note
Bump Golang to 1.23.1
```

### CODE REVIEW

**Comment by huww98** [comment]
BTW, do you think we should `export GOTOOLCHAIN=local` in run_with_go? If not, this function is meaningless because go will download the version specified in the `go.mod` itself.

**Comment by huww98** [comment]
/cc @xing-yang @jsafrane
PTAL

**Comment by jsafrane** [comment]
> BTW, do you think we should export GOTOOLCHAIN=local in run_with_go? If not, this function is meaningless because go will download the version specified in the go.mod itself. Or should we just remove run_with_go function and rely on go command to manage the version?

What approach does the main kubernetes/kubernetes use? Is there anything we could copy?

**Comment by pohly** [comment]
Manually downloading the right toolchain archive pre-dates the download mechanism in the Go toolchain itself. Nowadays using the toolchain to download the desired version is the right way to go.

**Comment by huww98** [comment]
Can we get this PR merged? Or should we fix the (seems unrelated) CI failure first?

**Comment by jsafrane** [comment]
We need the CI to be green before merge to make sure this PR does not break anything.

**Comment by jsafrane** [comment]
I am staring at the CI logs and they don't make much sense to me.

**Comment by pohly** [comment]
I suspect it fails because one or both repos are not completely present locally:
> --filter=blob:none will filter out all blobs (file contents) until needed by Git

This is used in https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/kubernetes-csi_csi-release-tools/261/pull-kubernetes-csi-release-tools-external-snapshotter/1854867176855441408/clone-log.txt.

I'm not sure yet how to change this - will ask on Slack.

**Comment by huww98** [comment]
> When cloning or fetching from a partial repository (i.e., one itself cloned with --filter), the server-side upload-pack may need to fetch extra objects from its upstream in order to complete the request. By default, upload-pack will refuse to perform such a lazy fetch

https://git-scm.com/docs/git-upload-pack

So I think set GIT_NO_LAZY_FETCH=0 should fix this.

File: pull-test.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@ CSI_RELEASE_TOOLS_DIR="$(pwd)"
 # Update the other repo.
 cd "$PULL_TEST_REPO_DIR"
 git reset --hard # Shouldn't be necessary, but somehow is to avoid "fatal: working tree has modifications.  Cannot add." (https://stackoverflow.com/questions/3623351/git-subtree-pull-says-that-the-working-tree-has-modifications-but-git-status-sa)
-git subtree pull --squash --prefix=release-tools "$CSI_RELEASE_TOOLS_DIR" master
+GIT_NO_LAZY_FETCH=0 git subtree pull --squash --prefix=release-tools "$CSI_RELEASE_TOOLS_DIR" master
```
Comment (jsafrane) [inline pull-test.sh:29]: Please add a comment why we need GIT_NO_LAZY_FETCH and perhaps put it to  a standalone line

**Comment by jsafrane** [comment]
Almost, please also revendor.

File: pull-test.sh (hunk 1 above)
Comment (huww98) [reply pull-test.sh]: Rebased onto #264

**Comment by huww98** [comment]
@pohly @jsafrane  Can you approve this PR? So that we can merge these two commits at once.

**Pull request merged**